banzai_nres/traces.py

```diff
@@ -0,0 +1,109 @@
+"""
+traces.py: Driver scripts for finding echelle orders across a CCD.
+
+Authors
+    G. Mirek Brandt (gmbrandt@ucsb.edu)
+"""
+
+from banzai_nres.utils.trace_utils import Trace, AllTraceFitter
+from banzai_nres.utils.db_utils import DataProduct
+from banzai.calibrations import CalibrationMaker, ApplyCalibration, create_master_calibration_header
+from banzai.utils import file_utils
+
+import banzai_nres.settings as nres_settings
+import banzai.settings as banzai_settings
+
+import sep
+import os
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class TraceMaker(CalibrationMaker):
+    def __init__(self, runtime_context):
+        super(TraceMaker, self).__init__(runtime_context)
+        self.runtime_context = runtime_context
+        self.order_of_poly_fit = nres_settings.TRACE_FIT_POLYNOMIAL_ORDER
+        self.second_order_coefficient_guess = nres_settings.TRACE_FIT_INITIAL_DEGREE_TWO_GUESS
+        self.trace_table_name = nres_settings.TRACE_TABLE_NAME
+        self.xmin = nres_settings.WINDOW_FOR_TRACE_IDENTIFICATION['min']
+        self.xmax = nres_settings.WINDOW_FOR_TRACE_IDENTIFICATION['max']
+        self.min_peak_to_peak_spacing = nres_settings.MIN_FIBER_TO_FIBER_SPACING
+        self.min_snr = nres_settings.MIN_SNR_FOR_TRACE_IDENTIFICATION
+
+    @property
+    def calibration_type(self):
+        return 'TRACE'
+
+    def make_master_calibration_frame(self, images):
+        traces = []
+        for image in images:
+            master_header = create_master_calibration_header(old_header=image.header, images=[image])
+            master_header['OBSTYPE'] = self.calibration_type
+            master_filename = banzai_settings.CALIBRATION_FILENAME_FUNCTIONS[self.calibration_type](image)
+            master_filepath = self._get_filepath(self.runtime_context, image, master_filename)
+            db_info = DataProduct(image=image)
+            db_info.is_master = True
```

Comment:
DataProduct is like an image but only with the relevant db_info from image that dbs.save_calibration_info needs (which it takes in the form of an image). This is a hotfix to avoid having to append the image onto the Trace object. However, this is gross. I would much rather update dbs.save_calibration_info upstream to take a single attribute of an image instead of the whole image object itself. That way I can simply append said attribute onto trace and avoid the DataProduct object.

Reply:
I would put all of these onto the trace object itself rather than creating a new data class.
        self.attributes = settings.CALIBRATION_SET_CRITERIA.get(self.obstype, {})

Reply:
done